library/core/src/any.rs

```diff
@@ -700,3 +770,284 @@ pub const fn type_name<T: ?Sized>() -> &'static str {
 pub const fn type_name_of_val<T: ?Sized>(_val: &T) -> &'static str {
     type_name::<T>()
 }
+
+///////////////////////////////////////////////////////////////////////////////
+// Provider trait
+///////////////////////////////////////////////////////////////////////////////
+
+/// Trait implemented by a type which can dynamically provide values based on type.
+#[unstable(feature = "provide_any", issue = "96024")]
+pub trait Provider {
+    /// Data providers should implement this method to provide *all* values they are able to
+    /// provide by using `demand`.
+    ///
+    /// # Examples
+    ///
+    /// Provides a reference to a field with type `String` as a `&str`.
+    ///
+    /// ```rust
+    /// # #![feature(provide_any)]
+    /// use std::any::{Provider, Demand};
+    /// # struct SomeConcreteType { field: String }
+    ///
+    /// impl Provider for SomeConcreteType {
+    ///     fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
+    ///         demand.provide_ref::<str>(&self.field);
+    ///     }
+    /// }
+    /// ```
+    #[unstable(feature = "provide_any", issue = "96024")]
+    fn provide<'a>(&'a self, demand: &mut Demand<'a>);
+}
+
+/// Request a value from the `Provider`.
+///
+/// # Examples
+///
+/// Get a string value from a provider.
+///
+/// ```rust
+/// # #![feature(provide_any)]
+/// use std::any::{Provider, request_value};
+///
+/// fn get_string<P: Provider>(provider: &P) -> String {
+///     request_value::<String, _>(provider).unwrap()
```

Comment:
not this PR: given that https://github.com/rust-lang/rust/issues/83701#issuecomment-1109958337 might stabilize, it might be interesting to turn `P` into an `impl Trait` so that `request_value::<String>(provider)` would just work here.

Reply:
Yes, I would like to - that is my motivation for trying to get 83701 stabilised :-)